design/pipeline-caching.md

```diff
@@ -0,0 +1,154 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a dependency caching mechanism that allows the outputs of steps within a pipeline to be optimized (or even skipped) if a suitable cached version of the outputs of those steps already exists. Caching can be effective provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch.
+
+## Proposal
+
+This proposal contains a number of interconnected elements. We have attempted to layer them in this document from the most primitive building blocks up to simplified YAML syntax to make caching "just work" in the majority of cases. Our approach to building a caching mechanism for Azure Pipelines will be to get the fundamental building blocks right and prove we can increase build performance and then evolve the YAML syntax once we know exactly what that syntax will need to describe.
+
+### Restore and Save Cache Tasks
+
+The basic functions of Pipeline Cache will be implemented by two tasks. The ```RestoreCache``` task and the ```SaveCache``` task. Both tasks will specify a collection of files and environment variables which can be used as a key to lookup the cache and identify a hit or a miss. Each task will also take a collection of paths which will be fetched from or stored in the cache associated with that key.
+
+Consider the following Azure Pipelines pipeline defined in YAML:
+
+```yaml
+steps:
+- script: yarn
+- script: yarn build
+```
+
+In this example we are working on a React-based application which was bootstrapped using ```create-react-app```. The output of this command is a directory with about ~28,000 files including the ```node_modules``` directory. On a local developer machine the ```yarn``` command will trigger the installation of any missing packages. Typically this will execute quickly because all packages already exist on the developer machine. However in cloud-based CI environments where build machines are recycled after every build there is no persistent state which means that Yarn must download all packages from scratch.
+
+Applying caching to this pipeline would involve adding a restore cache step and save cache step before and after the ```-script: yarn``` step.
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    key: |
+      package.json
+      yarn.lock
+    path: node_modules
+- script: yarn
+- task: SaveCache@0
+  inputs:
+    key: |
```

Comment:
Is the key required? In some cases (e.g. saving `ccache` or `clcache` contents), it shouldn't be needed, as the cache should be reusable for all similar jobs (`ccache` or `clcache` would disambiguate itself based on compiler version and options, AFAIU).

Reply:
I'm not sure if we explicitly guard against an empty key but if you just punched in a regular string we would interpret that as a constant. It might be better though to actually feed in variables that represent things such as a the compiler version and options for the sake of correctness.

That said for scenarios like ccache, clcache and sccache we are thinking about some kind of first class support for that. Its something that we'll start thinking about once we get past this initial building blocks phase and into righter integration with specific ecosystems.